crates/sargon/src/factor_instances_provider/factor_instances_cache/factor_instances_cache.rs

```diff
@@ -220,108 +225,284 @@ impl FactorInstancesCache {
-                            CACHE_FILLING_QUANTITY - count_in_cache;
-                        pf_pdp_qty_to_derive.append_or_insert_element_to(
-                            factor_source_id,
-                            (preset, qty_to_derive),
-                        );
-                    }
+                    let quantity_to_derive =
+                        cache_filling_quantity - count_in_cache;
+
+                    assert!(quantity_to_derive > 0);
+
+                    Some(CacheInstancesAndRemainingQuantityToDerive {
+                        instances_to_use_from_cache: FactorInstances::default(),
+                        quantity_to_derive,
+                    })
+                } else {
+                    None
+                };
+                if let Some(val) = val {
+                    per_factor_source.insert(*factor_source_id, val);
                 }
             }
+            if per_factor_source.is_empty() {
+                continue;
+            }
+            per_derivation_preset.insert(preset, per_factor_source);
         }
-        let outcome = if is_quantity_satisfied_for_all_factor_sources {
-            CachedInstancesWithQuantitiesOutcome::Satisfied(pf_instances)
+
+        let originally_request_presets = quantified_derivation_presets
+            .iter()
+            .map(|x| x.derivation_preset)
+            .collect::<IndexSet<_>>();
+
+        // The instances in the cache cannot satisfy the requested quantity
+        // we must derive more!
+        let is_quantity_unsatisfied_for_any_requested =
+            per_derivation_preset.iter().any(|(preset, pf)| {
+                originally_request_presets.contains(preset) /* Only lack of instances for originally requested presets is something which should cause the outcome of this reading from cache to be considered as `NotSatisfied` */
+                    && pf.iter().any(|(_, ci)| ci.quantity_to_derive > 0)
+            });
+
+        let outcome = if is_quantity_unsatisfied_for_any_requested {
+            CachedInstancesWithQuantitiesOutcome::NotSatisfied(
+                CacheNotSatisfied {
+                    cached_and_quantities_to_derive: per_derivation_preset,
+                },
+            )
         } else {
-            CachedInstancesWithQuantitiesOutcome::NotSatisfied {
-                partial_instances: pf_instances,
-                quantities_to_derive: pf_pdp_qty_to_derive,
-            }
+            CachedInstancesWithQuantitiesOutcome::Satisfied(CacheSatisfied {
+                cached: per_derivation_preset
+                    .into_iter()
+                    // Satisfied, but `per_derivation_preset` contains ALL Presets (in case of `NotSatisfied` - we are cache filling),
+                    // so we filter out only the originally requested ones.
+                    .filter(|(preset, _)| {
+                        originally_request_presets.contains(preset)
+                    })
```

Comment:
This filter could be done before `is_quantity_unsatisfied_for_any_requested` and then `originally_request_presets.contains(preset)` could be removed.

Reply:
you mean you can:

```rust
let per_derivation_preset = per_derivation_preset
                    .into_iter()
                    // Satisfied, but `per_derivation_preset` contains ALL Presets (in case of `NotSatisfied` - we are cache filling),
                    // so we filter out only the originally requested ones.
                    .filter(|(preset, _)| {
                        originally_request_presets.contains(preset)
                    }).collect();
```
?

That we cannot, since we need to *not* filter for case `CachedInstancesWithQuantitiesOutcome::NotSatisfied`.

So filtering should only happen for `CachedInstancesWithQuantitiesOutcome::Satisfied`.

Or maybe I misunderstood you..?

Reply:
sure, but you can have it stored in a separate variable, aka filtered_presets. This is a minor suggestion anyway.